README.md
# Simple Shadow Postcss Plugin

[![Build Status](https://travis-ci.org/mpeutz/postcss-simple-shadow.svg?branch=master)](https://travis-ci.org/mpeutz/postcss-simple-shadow)

Postcss function that will produce a Material design style drop shadow. There are 25 depths ranging from 0 25; 0 having no shadow while 25 has the most.

### Usage
---

Used as declaration is the quickest use of the shadow function. Adding an optional "color" with colorize the shadow. 

shadow: &lt;depth&gt; &lt;color (optional)&gt;;

- **Depth**: number value between 0 and 25.
- **Color**: valid css color - if left off the default #000 will be used.

```css
/* input.css */
shadow: 10;

/* output.css */
box-shadow: 0px 5px 7px -3px rgba(0, 0, 0, 0.2), 0px 10px 16px 2px rgba(0, 0, 0, 0.15), 0px 4px 20px 4px rgba(0, 0, 0, 0.09);

/* input.css */
shadow: 5 #00ddff;

/* output.css */
box-shadow: 0px 3px 4px -2px rgba(0, 221, 255, 0.15), 0px 5px 8px 1px rgba(0, 221, 255, 0.11), 0px 2px 10px 2px rgba(0, 221, 255, 0.07); 
```
---

Shadow can also be used as a property allowing it to be mixed with other shadow props. This can also take the optional color parameter.

box-shadow: &lt;additional properties ...&gt; shadow(&lt;depth&gt; &lt;color (optional)&gt;);

- **Depth**: number value between 0 and 25.
- **Color**: valid css color - if left off the default #000 will be used.

```css
/* input.css */
box-shadow: inset 0 1px 1px #000, shadow(15), inset 0 -1px 1px #fff;

/* output.css */
box-shadow: inset 0 1px 1px #000, 0px 7px 10px -5px rgba(0, 0, 0, 0.25), 0px 15px 24px 2px rgba(0, 0, 0, 0.18), 0px 6px 29px 5px rgba(0, 0, 0, 0.11), inset 0 -1px 1px #fff;

/* input.css */
box-shadow: shadow(15 #f00);

/* output.css */
box-shadow: 0px 7px 10px -5px rgba(255, 0, 0, 0.25), 0px 15px 24px 2px rgba(255, 0, 0, 0.18), 0px 6px 29px 5px rgba(255, 0, 0, 0.11);
```

ChangeLog.md
0.1.4 - 10 Mar 2019

* Added color parameter to make colored shadows.

0.1.3 - 04 Mar 2019

* Added inline shadow declaration

0.1.2 - 20 Feb 2019

* Updated Shadow Function to be more readable

0.1.1 - 17 Feb 2019

* Initial release

test/shadow.test.js
import postcss from 'postcss';
import test    from 'ava';

var plugin = require('../index.js');
function run(t, input, output, opts = { }) {
  return postcss([ plugin(opts) ]).process(input, {from: undefined})
    .then( result => {
      t.deepEqual(result.css, output);
      t.deepEqual(result.warnings().length, 0);
  });
}

test('handles shadows depth 0', t => {
  return run(t,
    '.shadow { shadow: 0; }',
    '.shadow { box-shadow: 0px 0px 0px -0px rgba(0, 0, 0, 0.1), 0px 0px 0px 0px rgba(0, 0, 0, 0.07), 0px 0px 0px 0px rgba(0, 0, 0, 0.05); }'
  );
});

test('handles shadows depth 5', t => {
    return run(t,
      '.shadow { shadow: 5; }',
      '.shadow { box-shadow: 0px 3px 4px -2px rgba(0, 0, 0, 0.15), 0px 5px 8px 1px rgba(0, 0, 0, 0.11), 0px 2px 10px 2px rgba(0, 0, 0, 0.07); }'
    );
});


test('handles shadows depth 10', t => {
  return run(t,
    '.shadow { shadow: 10; }',
    '.shadow { box-shadow: 0px 5px 7px -3px rgba(0, 0, 0, 0.2), 0px 10px 16px 2px rgba(0, 0, 0, 0.15), 0px 4px 20px 4px rgba(0, 0, 0, 0.09); }'
  );
});

test('handles shadows depth 15', t => {
  return run(t,
    '.shadow { shadow: 15; }',
    '.shadow { box-shadow: 0px 7px 10px -5px rgba(0, 0, 0, 0.25), 0px 15px 24px 2px rgba(0, 0, 0, 0.18), 0px 6px 29px 5px rgba(0, 0, 0, 0.11); }'
  );
});

test('handles shadows depth 20', t => {
  return run(t,
    '.shadow { shadow: 20; }',
    '.shadow { box-shadow: 0px 10px 13px -6px rgba(0, 0, 0, 0.3), 0px 20px 32px 3px rgba(0, 0, 0, 0.22), 0px 8px 39px 7px rgba(0, 0, 0, 0.13); }'
  );
});

test('handles shadows depth 25', t => {
  return run(t,
    '.shadow { shadow: 25; }',
    '.shadow { box-shadow: 0px 12px 16px -8px rgba(0, 0, 0, 0.35), 0px 25px 40px 4px rgba(0, 0, 0, 0.25), 0px 10px 48px 9px rgba(0, 0, 0, 0.15); }'
  );
});

test('handles inline shadows depth 25', t => {
  return run(t,
    '.shadow { box-shadow: 0 0 8px #000, shadow( 25), 0 0 8px #fff; }',
    '.shadow { box-shadow: 0 0 8px #000, 0px 12px 16px -8px rgba(0, 0, 0, 0.35), 0px 25px 40px 4px rgba(0, 0, 0, 0.25), 0px 10px 48px 9px rgba(0, 0, 0, 0.15), 0 0 8px #fff; }'
  );
});


test('handles shadows depth 5 with color', t => {
    return run(t,
      '.shadow { shadow: 5 #00ddff; }',
      '.shadow { box-shadow: 0px 3px 4px -2px rgba(0, 221, 255, 0.15), 0px 5px 8px 1px rgba(0, 221, 255, 0.11), 0px 2px 10px 2px rgba(0, 221, 255, 0.07); }'
    );
});

test('Handles inline shadow with color', t => {
  return run(t,
    '.shadow { box-shadow: shadow(15 #f00); }',
    '.shadow { box-shadow: 0px 7px 10px -5px rgba(255, 0, 0, 0.25), 0px 15px 24px 2px rgba(255, 0, 0, 0.18), 0px 6px 29px 5px rgba(255, 0, 0, 0.11); }'
  );
});

index.js
const postcss = require('postcss');
const valueParser = require('postcss-value-parser');
const tinycolor = require('tinycolor2');

const errorContext = {
	plugin: 'postcss-simple-shadow'
};

module.exports = postcss.plugin('shadow', () => {


    return function (css) {
        css.walkRules(rule => {

            rule.walkDecls(decl => {

                function shadows(depth, hue) {

                    if (depth < 0 || depth > 25) {
                        throw decl.error('shadow depth is out of range. [0-25]', errorContext);
                    }


                    const alpha = (num, in_min, in_max, out_min, out_max) => {
                        return Number(((num - in_min) * (out_max - out_min) / (in_max - in_min) + out_min).toFixed(2));
                    }

                    return `0px ${Math.ceil(depth * .458333)}px ${Math.ceil(depth * .625)}px -${Math.ceil(depth * .29166)}px ${tinycolor(hue).setAlpha(alpha(depth,0,25,0.1,0.35)).toRgbString()}, 0px ${Math.ceil(depth)}px ${Math.ceil(depth * 1.5833)}px ${Math.ceil(depth * .125)}px ${tinycolor(hue).setAlpha(alpha(depth,0,25,0.075,0.25)).toRgbString()}, 0px ${Math.ceil(depth * .375)}px ${Math.ceil(depth * 1.916)}px ${Math.ceil(depth * .3333)}px ${tinycolor(hue).setAlpha(alpha(depth,0,25,0.05,0.15)).toRgbString()}`;

                }

                function parseParams(val) {

                    let shadowParams = [];
                    

                    if (val.trim().indexOf(' ') != -1) {
                        shadowParams = val
                            .match(/^(\S+)\s(.*)/).slice(1)
                            .map(str => str.trim());
                    } else {
                        shadowParams = [val, '#000'];
                    }

                    if (! tinycolor(shadowParams[1]).isValid()) {
                        throw decl.error('Not a valid color', errorContext);
                    }


                    let dpth = parseInt(shadowParams[0]);
                    let col = shadowParams[1];

                    return [dpth, col]
                }


                if (decl.value.includes('shadow')) {

                    const parsedValue = valueParser(decl.value)

                    parsedValue.walk(node => {


                        if (node.value === 'shadow') {


                            let shade = parseParams(valueParser.stringify(node.nodes));
                            let rep = valueParser.stringify(node).replace('shadow', '');

                            node.value = shadows(shade[0], shade[1]);

                            decl.value = parsedValue.toString().replace(`${rep}`, '');
                        }
                    })

                }

                if (decl.prop == "shadow") {

                    let shade = parseParams(decl.value);

                    decl.cloneBefore({
                        prop: 'box-shadow',
                        value: shadows(shade[0], shade[1])
                    });

                    decl.remove();
                }


            })
        })

    };
});